text_encoder.go

```diff
@@ -0,0 +1,226 @@
+// RFC3339-formatted timestamps.
+func NewTextEncoder(options ...TextOption) Encoder {
+	enc := textPool.Get().(*textEncoder)
+	enc.truncate()
+	enc.timeFmt = time.RFC3339
+	for _, opt := range options {
+		opt.apply(enc)
+	}
+	return enc
+}
+
+func (enc *textEncoder) Free() {
+	textPool.Put(enc)
+}
+
+func (enc *textEncoder) AddString(key, val string) {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, val...)
+}
+
+func (enc *textEncoder) AddBool(key string, val bool) {
+	enc.addKey(key)
+	if val {
+		enc.bytes = append(enc.bytes, "true"...)
+		return
+	}
+	enc.bytes = append(enc.bytes, "false"...)
+}
+
+func (enc *textEncoder) AddInt(key string, val int) {
+	enc.AddInt64(key, int64(val))
+}
+
+func (enc *textEncoder) AddInt64(key string, val int64) {
+	enc.addKey(key)
+	enc.bytes = strconv.AppendInt(enc.bytes, val, 10)
+}
+
+// TODO: uncomment once #116 lands.
+// func (enc *textEncoder) AddUint(key string, val uint) {
+// 	enc.AddUint64(key, uint64(val))
+// }
+
+// func (enc *textEncoder) AddUint64(key string, val uint64) {
+// 	enc.addKey(key)
+// 	enc.bytes = strconv.AppendUint(enc.bytes, val, 10)
+// }
+
+func (enc *textEncoder) AddFloat64(key string, val float64) {
+	enc.addKey(key)
+	switch {
+	case math.IsNaN(val):
+		enc.bytes = append(enc.bytes, "NaN"...)
+	case math.IsInf(val, 1):
+		enc.bytes = append(enc.bytes, "+Inf"...)
+	case math.IsInf(val, -1):
+		enc.bytes = append(enc.bytes, "-Inf"...)
+	default:
+		enc.bytes = strconv.AppendFloat(enc.bytes, val, 'f', -1, 64)
+	}
+}
+
+func (enc *textEncoder) AddMarshaler(key string, obj LogMarshaler) error {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, '{')
+	err := obj.MarshalLog(enc)
+	enc.bytes = append(enc.bytes, '}')
+	return err
+}
+
+func (enc *textEncoder) AddObject(key string, obj interface{}) error {
+	enc.AddString(key, fmt.Sprintf("%+v", obj))
+	return nil
+}
+
+func (enc *textEncoder) Clone() Encoder {
+	clone := textPool.Get().(*textEncoder)
+	clone.truncate()
+	clone.bytes = append(clone.bytes, enc.bytes...)
+	clone.timeFmt = enc.timeFmt
+	return clone
+}
+
+func (enc *textEncoder) WriteEntry(sink io.Writer, msg string, lvl Level, t time.Time) error {
+	if sink == nil {
+		return errNilSink
+	}
+
+	final := textPool.Get().(*textEncoder)
+	final.truncate()
+	enc.addLevel(final, lvl)
+	enc.addTime(final, t)
+	enc.addMessage(final, msg)
+
+	if len(enc.bytes) > 0 {
+		final.bytes = append(final.bytes, ' ')
+		final.bytes = append(final.bytes, enc.bytes...)
+	}
+	final.bytes = append(final.bytes, '\n')
+
+	expectedBytes := len(final.bytes)
+	n, err := sink.Write(final.bytes)
+	final.Free()
+	if err != nil {
+		return err
+	}
+	if n != expectedBytes {
+		return fmt.Errorf("incomplete write: only wrote %v of %v bytes", n, expectedBytes)
+	}
+	return nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.bytes = enc.bytes[:0]
+}
+
+func (enc *textEncoder) addKey(key string) {
+	lastIdx := len(enc.bytes) - 1
+	if lastIdx >= 0 && enc.bytes[lastIdx] != '{' {
```

Comment:
This seems hacky and incorrect.

E.g., if I do `addString("key", "{")` then the value in `bytes` is `key={` and this will check the last byte, see that it's `{`, and drop the space.

Use a separate bool to track state, and suggest adding failing tests as well.